cluster-autoscaler/core/scaleup/orchestrator/orchestrator.go

```diff
@@ -119,32 +119,53 @@ func (o *ScaleUpOrchestrator) ScaleUp(
 		}
 	}
 
+	var nodeGroupPoolList [][]cloudprovider.NodeGroup
+	// Add cloudprovider nodegroup pool into the list.
+	nodeGroupPoolList = append(nodeGroupPoolList, nodeGroups)
+
+	if o.processors != nil && o.processors.NodeGroupPoolListProcessor != nil {
+		var err error
+		nodeGroupPoolList, nodeInfos, err = o.processors.NodeGroupPoolListProcessor.Process(o.autoscalingContext, nodeGroupPoolList, nodeInfos, unschedulablePods)
+		if err != nil {
+			return scaleUpError(&status.ScaleUpStatus{}, errors.ToAutoscalerError(errors.InternalError, err))
+		}
+	}
+
 	resourcesLeft, aErr := o.resourceManager.ResourcesLeft(o.autoscalingContext, nodeInfos, nodes)
 	if aErr != nil {
 		return scaleUpError(&status.ScaleUpStatus{}, aErr.AddPrefix("could not compute total resources: "))
 	}
 
 	now := time.Now()
 
-	// Filter out invalid node groups
-	validNodeGroups, skippedNodeGroups := o.filterValidScaleUpNodeGroups(nodeGroups, nodeInfos, resourcesLeft, now)
-
-	// Calculate expansion options
-	schedulablePods := map[string][]*apiv1.Pod{}
+	var validNodeGroups []cloudprovider.NodeGroup
+	var skippedNodeGroups map[string]status.Reasons
 	var options []expander.Option
+	schedulablePods := map[string][]*apiv1.Pod{}
 
-	for _, nodeGroup := range validNodeGroups {
-		schedulablePods[nodeGroup.Id()] = o.SchedulablePods(podEquivalenceGroups, nodeGroup, nodeInfos[nodeGroup.Id()])
-	}
+	for _, nodeGroups = range nodeGroupPoolList {
+		// Filter out invalid node groups
+		validNodeGroups, skippedNodeGroups = o.filterValidScaleUpNodeGroups(nodeGroups, nodeInfos, resourcesLeft, now)
+		for _, nodeGroup := range validNodeGroups {
+			schedulablePods[nodeGroup.Id()] = o.SchedulablePods(podEquivalenceGroups, nodeGroup, nodeInfos[nodeGroup.Id()])
+		}
 
-	for _, nodeGroup := range validNodeGroups {
-		option := o.ComputeExpansionOption(nodeGroup, schedulablePods, nodeInfos, upcomingNodes, now)
-		if len(option.Pods) == 0 || option.NodeCount == 0 {
-			klog.V(4).Infof("No pod can fit to %s", nodeGroup.Id())
-			continue
+		for _, nodeGroup := range validNodeGroups {
+			option := o.ComputeExpansionOption(nodeGroup, schedulablePods, nodeInfos, upcomingNodes, now)
+			if len(option.Pods) == 0 || option.NodeCount == 0 {
+				klog.V(4).Infof("No pod can fit to %s", nodeGroup.Id())
+				continue
+			}
+
+			options = append(options, option)
 		}
 
-		options = append(options, option)
+		// Stop binpacking early if some pool has expansion options.
+		if len(options) > 0 {
+			break
+		}
+		// No expansion options found in this pool. Resetting map.
+		schedulablePods = map[string][]*apiv1.Pod{}
```

Comment:
This is not strictly necessary, I think that leaving these cached schedulablePods should not break anything.

Reply:
Yes, was just saving some space. Anyway changed.